src/Linters/ShoutCaseEnumMembersLinter.hack

```diff
@@ -0,0 +1,129 @@
+ */
+
+namespace Facebook\HHAST;
+
+use namespace HH\Lib\{C, Dict, Regex, Str, Vec};
+
+final class ShoutCaseEnumMembersLinter extends AutoFixingASTLinter {
+  const type TContext = EnumDeclaration;
+  const type TNode = Enumerator;
+
+  /* @var enum_name -> shout_name to original_name*/
+  private static dict<string, dict<string, string>> $newNames = dict[];
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    self::TContext $enum,
+    self::TNode $member,
+  ): ?ASTLintError {
+    $original_name = self::memberToName($member);
+    if (self::isShoutCase($original_name)) {
+      return null;
+    }
+    self::prepareAllNewNamesFor($enum);
+    $enum_name = $enum->getName()->getText();
+    $shout_name = self::transformToShoutCase($original_name);
+
+    // If we can't find the shout name, we bailed, because the enum
+    // member names were not unique.
+    // If the shout name maps to something other than the orignal,
+    // we would cause a collision.
+    // We bail out either way.
+    if (
+      !C\contains_key(self::$newNames[$enum_name], $shout_name) ||
+      self::$newNames[$enum_name][$shout_name] !== $original_name
+    ) {
+      return new ASTLintError(
+        $this,
+        Str\format(
+          'Member {%s} is not in SHOUT_CASE,
+but the autogenerated name collides or your enum contains duplicate names,
+so no autofix is suggested.',
+          $original_name,
+        ),
+        $member,
+        null,
+      );
+    }
+
+    return new ASTLintError(
+      $this,
+      Str\format('Member {%s} is not in SHOUT_CASE', $original_name),
+      $member,
+      () ==> self::getFixedNode($member),
+    );
+  }
+
+  <<__Memoize>>
+  private static function isShoutCase(string $member_name): bool {
+    return Regex\matches($member_name, re"/^[A-Z0-9_]+$/");
+  }
+
+  private static function memberToName(Enumerator $member): string {
+    return $member->getName()->getLastTokenx()->getText();
+  }
+
+  private static function prepareAllNewNamesFor(EnumDeclaration $enum): void {
+    $enum_name = $enum->getName()->getText();
+    if (C\contains_key(self::$newNames, $enum_name)) {
+      return;
+    }
+
+    $old_names = Vec\map(
+      $enum->getChildren()['enumerators']->toVec(),
+      $member ==> self::memberToName($member as Enumerator),
+    );
+    $old_names_unique = keyset($old_names);
+
+    // !C\all_unique() -> bail
+    if (C\count($old_names_unique) !== C\count($old_names)) {
+      self::$newNames[$enum_name] = dict[];
+      return;
+    }
+
+    // Put all non-shout case names in front.
+    // This way they will be overwritten if there is a shout case member
+    // that has the same name as the suggested fix.
+    $old_names_unique = Vec\partition(
+      $old_names_unique,
+      $name ==> self::isShoutCase($name),
+    )
+      |> Vec\concat($$[1], $$[0]);
+
+    self::$newNames[$enum_name] = Dict\from_values(
+      $old_names_unique,
+      $name ==> self::transformToShoutCase($name),
+    );
```

Comment:
If you `Dict\flip` this at the end (so it's a map from old name => new name), it should simplify the code in `getLintErrorForNode` (and potentially improve performance), since you won't have to transform each value to shoutcase again, it would just be a simple dictionary lookup.

Reply:
Ah, that is indeed a smart one. I needed the shot names to be the keys for once because this removes duplicates via key collision. However, flipping it afterwards is perfectly legal. Will do tomorrow.

Reply:
> I needed the shot names to be the keys for once because this removes duplicates via key collision.

Yeah that was a nice trick. It would have taken me a long time to figure out what was going on if it weren't for the comment, so thanks for adding that :)